src/JsonApiSerializer/JsonConverters/ResourceObjectConverter.cs

```diff
@@ -149,146 +167,238 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
             }
         }
 
-
-        protected void WriteFullObjectJson(JsonWriter writer, object value, JsonSerializer serializer)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteFullObjectJson(JsonWriter writer, object value, JsonSerializer serializer)
         {
             var valueType = value.GetType();
             var contractResolver = serializer.ContractResolver;
 
             if (!(contractResolver.ResolveContract(valueType) is JsonObjectContract contract
                  && contract.Converter is ResourceObjectConverter))
-                throw new JsonApiFormatException(writer.Path, 
+                throw new JsonApiFormatException(writer.Path,
                     $"Expected to find to find resource object, but found '{value}'",
-                    "Resource indentifier objects MUST contain 'id' members");
+                    "Resource identifier objects MUST contain 'id' members");
 
-            //prepare to start capturing attributes and relationships
-            var attributes = new List<KeyValuePair<JsonProperty, object>>();
-            var relationships = new List<KeyValuePair<JsonProperty, object>>();
+            // prepare to start capturing attributes and relationships
 
-            //will capture id and type as we go through
-            object id = null;
-            object type = null;
+            // default list size to the number of expected properties 
+            // to reduce allocations
+            List<KeyValuePair<string, object>> relationships = null;
 
             writer.WriteStartObject();
 
-            //A resource object MUST contain at least the following top-level members: type
-            var typeProp = contract.Properties.GetClosestMatchProperty("type");
-            if (typeProp == null)
+            // certain properties are written outside of the attributes block
+            // we handle these first so that we can make a single pass through the
+            // properties collection
+
+            // id is optional
+            string id = null;
+            var idProperty = contract.Properties.GetClosestMatchProperty(nameof(PropertyNames.Id));

```

Comment:
dont use `nameof` - it is misleading to use the property name rather than the value

The speedup is probably becuase it does a case sensitive check before a case insensitive, and most C# properties will have an uppercase. Probably clearer to create some more consts

Reply:
I can also use the existing property values and use GetProperty with a case insensitive search. This is actually quicker for the test case but as it iterates the object properties I suspect it’s only faster if the object is small. 

I’ll add some new constants - I did that originally.